mixnet/fisheryates.py

```diff
@@ -0,0 +1,15 @@
+from typing import List
+
+from numpy.random import Generator
+from randomgen.chacha import ChaCha
+
+
+class FisherYates:
+    @staticmethod
+    def shuffle(elements: List, entropy: bytes) -> List:
+        out = elements.copy()
+        rng = Generator(ChaCha(seed=list(entropy)))
+        for i in reversed(range(1, len(out))):
+            j = rng.integers(low=0, high=i, endpoint=True)  # [low,high]
+            out[i], out[j] = out[j], out[i]
+        return out
```

Comment:
Fisher yates is the default python shuffler btw. Probably not worth to maintain it in specs and would be simplers. We have a wrapper in the carnot package you can take a look at (We can copy paste that instead, better not to have inter package dependencies).

Reply:
Maybe something like?:

```python
class FisherYates:
    @staticmethod
    def shuffle(elements: List, entropy: bytes) -> List:
        """
       Fisher-yates shuffling algorithm
       It is the one used by python by default
       https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle
       https://softwareengineering.stackexchange.com/a/215780
       :param elements:
       :param entropy:
       :return:
       """
        out = elements.copy()
        random.seed(a=entropy, version=2)
        random.shuffle(out)
        # reset seed
        random.seed()
        return out
```

Reply:
Oh, I didn't know that the `random.shuffle` is actually the Fisher-Yates.

Also, I found that we can use `random.sample` instead of using `random.shuffle` and implementing sampling by ourselves, as described in https://docs.python.org/3/library/random.html#random.shuffle. I don't think this abstraction doesn't reduce readability of the specification. 
705dfeb994ef2e08e1e6b4927dc3b3f73fcb6ca8

Reply:
Yes, `random.sample` can be used. But we need to know what sampling algorithm it uses as it needs to be clear in the specification.

Reply:
Oh, just saw in the commit that sampel it actually uses *fisher_yates* internally. Neat!

```
A sampling algorithm using Fisher-Yates shuffling.
        In Python, random.sample uses Fisher-Yates shuffling internally by default.
        https://softwareengineering.stackexchange.com/a/215780
        https://docs.python.org/3/library/random.html#random.shuffle
```

Reply:
Ah, I just looked at the internal source code of `random.sample`. It seems that it doesn't use FisherYates sampling internally. I had guessed it from the [sentence](https://docs.python.org/3/library/random.html#random.shuffle): 
>  random.shuffle(x)
>
>    Shuffle the sequence x in place.
>
>    To shuffle an immutable sequence and return a new shuffled list, use sample(x, k=len(x)) instead.

, but I think it's better to use `random.shuffle` explicitly and implement sampling as I implemented at the beginning.